[JEP-237] Introduce initial documentation for FIPS-140 (#6675)
jenkins-infra/jenkins.io: content/doc/book/system-administration/FIPS-140.adoc

---
layout: section
---
ifdef::backend-html5[]
ifndef::env-github[:imagesdir: ../../resources/managing]
:notitle:
:description:
:author:
:email: jenkinsci-users@googlegroups.com
:sectanchors:
:toc: left
endif::[]

= FIPS-140

[IMPORTANT]

> ----

Author: ```suggestion
====
```

> support blurb

Author: need to state that this is for informational purposes, the Jenkins project does not claim that Community does actively check for FIPS compliance in Jenkins or its plugins.

> ----

Author: ```suggestion
====
```

== Jenkins 

> It may be possible to run Jenkins in a FIPS-140 compliant nature when the <<managing/system-properties#jenkins-security-FIPS140-COMPLIANCE, compliance flag>> is enabled, the servlet container is appropriately configured, the JVM is appropriately configured, and the host OS is appropriately configured.

Reviewer: ```suggestion
It may be possible to run Jenkins in a FIPS-140 compliant environment when the <<managing/system-properties#jenkins-security-FIPS140-COMPLIANCE, compliance flag>> is enabled, and the servlet container, the JVM, and the host OS are all appropriately configured.
```

Author: it is not the environment - you can likely run Jenkins in a FIPS compliant environment today - but Jenkins itself will not be FIPS compliant if you do that as it uses non approved encryption libraries.

perhaps `way` is better than `nature`?

```suggestion
It may be possible to run Jenkins in a FIPS-140 compliant way when the <<managing/system-properties#jenkins-security-FIPS140-COMPLIANCE, compliance flag>> is enabled, and the servlet container, the JVM, and the host OS are all appropriately configured.```

Author: ```suggestion
It may be possible to run Jenkins in a FIPS-140 compliant manner when the <<managing/system-properties#jenkins-security-FIPS140-COMPLIANCE, compliance flag>> is enabled, and the servlet container, the JVM, and the host OS are all appropriately configured.
```

> It may be possible to run Jenkins in a FIPS-140 compliant manner when the <<managing/system-properties#jenkins-security-FIPS140-COMPLIANCE, compliance flag>> is enabled, and the servlet container, the JVM, and the host OS are all appropriately configured.

Author: ```suggestion
It may be possible to run Jenkins in a FIPS-140 compliant manner when the <<../managing/system-properties#jenkins-security-fips140-compliance, compliance flag>> is enabled, and the servlet container, the JVM, and the host OS are all appropriately configured.
```

How to configure the servlet container, JVM and host are out of scope of the Jenkins community project as this is a complex area with many pitfalls and gotchas.
Some Jenkins features may not work or be disabled.

== Plugins

> Plugins may or may not honour a request to run in FIPS-140 compliance mode.

Reviewer: ```suggestion
Plugins may or may not honor a request to run in FIPS-140 compliance mode.
```

Author: I am happy with the Kings English here :)

> Before you install or upgrade any plugin, you should make checks of the plugin's code to ensure it adheres to the FIPS-140 standard.

Reviewer 2: ```suggestion
Before you install or upgrade any plugin, you check the plugin's code to ensure it adheres to the FIPS-140 standard.
```

> As such before installing or upgrading any plugin you should make checks of the plugins code to ensure it adheres to the FIPS-140 standard.

Reviewer: ```suggestion
Before you install or upgrade any plugin, you should make checks of the plugin's code to ensure it adheres to the FIPS-140 standard.
```

== What FIPS-140 mode does

> Enabling <<managing/system-properties#jenkins-security-FIPS140-COMPLIANCE, FIPS mode>> provides a hint to Jenkins and any plugins that have opted in that they should prefer cryptographic algorithms that *may*.footnote:[Algorithms are not approved, rather an implementation of an algorithm is.  

Reviewer: ```suggestion
If you enable <<managing/system-properties#jenkins-security-FIPS140-COMPLIANCE, FIPS mode>>, it provides a hint to Jenkins and any plugins that have opted in that they should prefer cryptographic algorithms that *may* enable FIPS-140 mode.footnote:[Algorithms are not approved, rather an implementation of an algorithm is.  
```

Author: incorporated sections of the comment.

> If you enable <<managing/system-properties#jenkins-security-FIPS140-COMPLIANCE, FIPS-140 mode>>, it provides a hint to Jenkins and any plugins that have opted in that they should prefer cryptographic algorithms that *may*.footnote:[Algorithms are not approved, rather a specific implementation of specific  algorithm is.  

Author: ```suggestion
If you enable <<../managing/system-properties#jenkins-security-fips140-compliance, FIPS-140 mode>>, it provides a hint to Jenkins and any plugins that have opted in that they should prefer cryptographic algorithms that *may*.footnote:[Algorithms are not approved, rather a specific implementation of specific  algorithm is.  
```

> However the implementation used at runtime will depend on the JVM, JVM configuration, as well as the host OS.

Reviewer: ```suggestion
However, the implementation used at runtime depends on the JVM, JVM configuration, and the host OS.
```

> As this is outside the scope of the Jenkins project the algorithms targeted are known to be available from at least one link:https://csrc.nist.gov/projects/cryptographic-module-validation-program/validated-modules/search[FIPS compliant provider], namely the link:https://csrc.nist.gov/projects/cryptographic-module-validation-program/certificate/3514[BouncyCastle FIPS library].] be FIPS approved.

Reviewer: ```suggestion
As this is outside the scope of the Jenkins project, the algorithms targeted are available from at least one link:https://csrc.nist.gov/projects/cryptographic-module-validation-program/validated-modules/search[FIPS compliant provider], namely the link:https://csrc.nist.gov/projects/cryptographic-module-validation-program/certificate/3514[BouncyCastle FIPS library].] that is FIPS approved.
```

Author: this and some preceeding no not make sense.  "that is FIPS approved" has now become a dangling sentence.

"if you enable <<managing/system-properties#jenkins-security-FIPS140-COMPLIANCE, FIPS mode>>, it provides a hint to Jenkins and any plugins that have opted in that they should prefer cryptographic algorithms that *may* enable FIPS-140 mode[1]. that is FIPS approved"

> This may mean that some features are disabled entirely, or even may use a less secure (but compliant) form of cryptography than normal.

Reviewer 2: ```suggestion
This may mean that some features are disabled entirely, or may use a less secure (but compliant) form of cryptography than normal.
```

== What FIPS-140 mode does not do

> FIPS mode is not a magic bullet, if any code from the JVM, servlet container, Jenkins or any plugin asks for a non compliant algorithm this will still be the case, and the request may well be honoured.

Reviewer 2: Are we going with "non compliant" or "non-compliant" _(with a hyphen)_? Need to check for consistency 🙂

> FIPS mode is not a magic bullet, if any code from the JVM, servlet container, Jenkins or any plugin asks for a non compliant algorithm this will still be the case, and the request may well be honoured.

Reviewer: ```suggestion
If any code from the JVM, servlet container, Jenkins, or any plugin requests a noncompliant algorithm, this will still be the case, and the request may be honored.
```

Author: ```suggestion
If any code from the JVM, servlet container, Jenkins, or any plugin requests a non-compliant algorithm, this will still be the case, and the request may be honoured.
```

> For example this mode does not and can not configure the JVM, so TLS connections to external secure web sites may still use non compliant cryptography.

Reviewer: ```suggestion
For example, this mode cannot configure the JVM, so TLS connections to external secure web sites might still use noncompliant cryptography.
```

Author: ```suggestion
For example, this mode cannot configure the JVM, so TLS connections to external secure web sites might still use non-compliant cryptography.
```

> Further, Jenkins cannot ensure that plugins even use encryption at all when appropriate.

Reviewer: ```suggestion
Additionally, Jenkins cannot ensure that plugins will even use encryption at all, when appropriate.
```

At the end of the day, just because Jenkins and plugins run when FIPS mode is enabled does not mean that it adheres to the USA government standard.

== How to run a fully FIPS-140 compliant Jenkins

> As mentioned above the host, JVM and servlet container all need to be configured appropriately to ensure Jenkins is FIPS-140 compliant.  

Reviewer: ```suggestion
As previously mentioned, the host, JVM, and the servlet container all need to be configured appropriately to ensure that Jenkins is FIPS-140 compliant.  
```

> Additionally extreme care should be taken when installing or upgrading plugins as they may or may not be FIPS-140 compliant, and they may introduce code that is non compliant or otherwise change the JVM configuration in such a way that breaks compliance.

Reviewer: ```suggestion
Extreme care should be taken when installing or upgrading plugins as they may or may not be FIPS-140 compliant, and they may introduce code that is noncompliant or otherwise change the JVM configuration iso that it breaks compliance.
```

Author: ```suggestion
Extreme care should be taken when installing or upgrading plugins as they may or may not be FIPS-140 compliant, and they may introduce code that is non-compliant or otherwise change the JVM configuration so that it breaks compliance.
```

> The Jenkins community does not support Jenkins FIPS mode, and due to the complex nature of JVM and servlet configuration that can change between versions, will not provide documentation for this.

Reviewer: ```suggestion
The Jenkins community does not support Jenkins FIPS mode, and due to the complex nature of JVM and servlet configuration that can change between versions, they will not provide documentation for it.
```

Author: `they` seems wrong - this is community documentation written and hosted by the comminuty so "they" feels incorrect to me here?  Could be "we" but to some audiences when reading could be partially confusing?

Reviewer 2: I think defining who it is would be better e.g. 'the Jenkins community will not provide documentation'

Author: sentence has been reworked

> Should you wish to run Jenkins in a FIPS compliant way it is therefore recommended to obtain support from a commercial vendor that will support you in doing so.

Reviewer: ```suggestion
If you want to run Jenkins in a FIPS-compliant environment, it is recommended that you obtain support from a commercial vendor.
```

Author: it is not the environment that you want to run Jenkins in - it is that you want to run Jenkins FIPS-140  Compliant.

```suggestion
If you need to run Jenkins in a way that it is FIPS compliant, it is recommended that you obtain support from a commercial vendor.
```

> However the Jenkins community may take steps to fix issues relating to FIPS compliance, these will be treated as any other bug report.

Reviewer: ```suggestion
The Jenkins community may be able to fix issues relating to FIPS compliance; these will be treated as any other bug report.
```

Author: ```suggestion
The Jenkins community may be able to fix issues relating to FIPS compliance; these will be treated as any other bug report or feature request.
```
